doc/specs/#1595 - Suggestions UI/Snippets.md

```diff
@@ -0,0 +1,526 @@
+            "  $resp = Add-GitHubBetaProjectItem -ProjectNodeId $s.organization.projectV2.id -ContentNodeId $_ ;",
+            "}",
+            ""
+        ]
+    },
+    "name": "Upload to project board",
+    "description": "Sync all our issues and bugs that have been triaged and are actually on the backlog to the big-ol project",
+},
+```
+
+This is slightly more maintainable. But now, a user could set
+`"waitForSuccess": true`, and if any part of the script fails, then the rest of
+it won't be sent to the shell.
+
+#### Fragment actions
+
+This was already added in [#16185]. These will allow third-party developers to
+create apps which add additional snippets to the Terminal. These will require
+app developers to add `id`s to each action they add in this way. Users can then
+bind that action `id` to a keybinding, if they so choose.
+
+Case in point:
+https://github.com/abduvik/just-enough-series/tree/master/courses/docker+docker-compose.
+Something like that should be able to be easily added directly to the Terminal.
+
+### Snippets pane
+
+With non-terminal content landing in 1.21 Preview, it's now simple to add
+additional types of panes to add to the Terminal. We'll support a new pane
+`"type": "snippets"`, to support opening the Snippets pane.
+
+This will be a pane with a `TreeView` in it and a text box to filter results
+(ala the Command Palette).
+
+Each item in the TreeView will be a kind of `FilteredCommand`, with a play
+button to support quickly running the command.
+
+This pane could also support all the different suggestion sources that the
+Suggestions UI supports - `recentCommands` could be plumbed into it from the
+currently active This pane could also support checkboxes to filter different
+suggestion sources.
+
+### Per-Project Snippets (`.wt.json`)
+
+> [INFO!]
+>
+> TODO!: Let's make sure to discuss the filename. It doesn't need to be
+> `.wt.json`. That seemed to match things like `.clang-format`, `.vsconfig`,
+> etc, but then also included the extension. However, node projects just use
+> `package.json` without the leading `.` for storing per-project commands.
+> Perhaps `.wt.json` is the worst of both? The best of both? `.wtconfig`?
+
+Users may also want to leave snippets in the root of their repo, for others to
+use as well. To support this, the Terminal will automatically look for a
+`.wt.json` file in any directories that are parents of the CWD of the shell, and
+load actions from that file as if it were a fragment extension as well. That
+will start with the `startingDirectory` for any new panes created. If the user
+has shell integration configured to tell the Terminal about the CWD, then we'll
+refresh that list as the user changes directories.
+
+* In `Terminal.Settings.Model`, we will store a cached map of path->actions.
+  * that if multiple panes are all in the same CWD, they don't need to
+    individually re-read the file from disk and regenerate that part of the map.

```

Comment:
In my current implementation, I don't hot-reload that file. I'll warm-reload it (if you update the settings, we'll flush the cached actions and re-load them). 

* From purely a theoretical standpoint, I don't think we should try to read the file from disk literally every time the user asks for suggestions. On paper that sounds bad
* From just local testing, reading the file from disk every time was totally fine. 
* We could maybe optimize this a little by also caching the last `LastWriteTime` of the file, and only loading it if the timestamp is different. 🤷

Reply:
(Circling back to this) I'm thinking about the scenario where the .wt.json was improperly written. We'll try to load it then display some kind of warning to the user. Let's assume the user/author then goes into the .wt.json to fix it. We'll probably want to reload it. Otherwise, the dev loop for that would require the user to (1) know that they have to update their settings to re-load the .wt.json or (2) be a bit confused about why their changes to .wt.json aren't being propagated.